fix: prevent duplicate error toasts and improve error message for appointment creation

harsh04044 opened the pull request:
## Proposed Changes

Fixes #14954 

This PR fixes the issue where patient registration showed a success message before appointment creation completed. Previously, when appointment creation failed, users would still see the "Patient created successfully" toast, which was misleading since the overall registration process (patient + appointment) had not completed successfully.

**Changes:**
- Added `onError` handler to the `createAppointment` mutation that displays a user-friendly error message ("Failed to book appointment") and redirects users to `/patient/home` when appointment creation fails
- Added `silent: true` to the `createAppointment` mutation to prevent global error handler interference
- Fixed network error handling in `query.ts` to properly respect the `silent` option for network errors
- Added translation key `failed_to_create_appointment` with the message "Failed to book appointment" for better user experience

**Before**

https://github.com/user-attachments/assets/c23216e8-59ce-4d10-8cfe-7e56377c0ddc

**After**

https://github.com/user-attachments/assets/36bd5c60-a4c1-407c-b654-edbc5098403a



**Context:**
Now when appointment creation fails, users see:
1. "Patient created successfully" toast (correct - patient was created)
2. "Failed to book appointment" toast (clear error message)
3. User is redirected to `/patient/home`

This ensures users understand that while their patient record was created, the appointment booking failed and they need to try again.

Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate the bug or test the new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is placed in I18n files.
- [ ] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [x] Request peer reviews.
- [ ] Complete QA on mobile devices.
- [x] Complete QA on desktop devices.
- [ ] Add or update Playwright tests for related changes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Improvements**
  * Enhanced error handling for appointment creation. Users will now receive a failure notification when appointment creation fails and be automatically redirected to the patient home page.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

harsh04044 committed:
fix: prevent duplicate error toasts and improve error message for appointment creation

- Add silent: true to createAppointment mutation to prevent global error handler from showing duplicate toasts
- Add onError handler with user-friendly error message and redirect
- Fix network error handling in query.ts to respect silent option
- Add translation key 'failed_to_create_appointment' with message 'Failed to book appointment'

Fixes #14954

harsh04044 committed:
refactor: address CodeRabbit suggestions for type safety and code quality

- Add NetworkError class to replace unsafe type assertions in query.ts
- Fix network error handling to respect silent option without fake Response
- Add error logging in PatientRegistration for debugging
- Refactor nested ternary in UserForm to computed buttonLabel constant

All changes improve type safety, maintainability, and debugging capabilities.

harsh04044 committed:
refactor: make NetworkError constructor more flexible

- Make silent parameter optional with default false for better ergonomics
- Add optional cause parameter for error chaining (consistent with HTTPError)
- Pass cause to super() for better diagnostics

Addresses CodeRabbit suggestion for improved flexibility and consistency.

SiddanthNayak reviewed (commented):
This does not fix the issue though? 

Why is onSuccess getting executed when the request is not successfull?

Jacobjeevan commented on src/components/Users/UserForm.tsx:
unrelated changes

Jacobjeevan commented on src/pages/PublicAppointments/PatientRegistration.tsx:
```diff
@@ -108,6 +108,7 @@ export function PatientRegistration(props: PatientRegistrationProps) {
         headers: {
           Authorization: `Bearer ${tokenData.token}`,
         },
+        silent: true,
```
let's not make this query silent

Jacobjeevan commented on src/pages/PublicAppointments/PatientRegistration.tsx:
```diff
@@ -124,6 +125,11 @@ export function PatientRegistration(props: PatientRegistrationProps) {
         );
       },
+      onError: (error) => {
+        console.error("Appointment creation failed:", error);
```
discard console logging

harsh04044 replied:
@Jacobjeevan 
i have  done it only cause of coderabbit suggestion i will change this in my latest commit

Jacobjeevan commented on src/Utils/request/query.ts:
discard these changes

Jacobjeevan commented on src/Utils/request/types.ts:
discard

harsh04044 committed:
revert: remove unrelated NetworkError changes and restore original translation

- Remove NetworkError class from types.ts (unrelated to issue #14954)
- Revert query.ts to original error handling (no NetworkError usage)
- Restore original translation 'Failed to create an appointment'

Addresses maintainer feedback to keep PR focused on issue #14954 only.

harsh04044 commented:
Hi @Jacobjeevan , i have question about error handling.

After removing `silent: true`, we're getting duplicate error toasts:
1. Global handler shows "Network Error"
2. Our `onError` shows "Failed to create appointment"

i have thought of this approach: Using `toast.dismiss()` in `onError` to dismiss the global toast, then showing our custom message.

Is this acceptable, or is there a preferred pattern for custom error messages without duplicates?

onError: (_error) => {
  toast.dismiss(); // Dismiss global error handler toast
  toast.error(t("failed_to_create_appointment"));
  navigate("/patient/home", { replace: true });
},

And i have addressed all the changes u suggested i will commit them with this one after your suggestion

Jacobjeevan commented:
> Hi @Jacobjeevan , i have question about error handling.
> 
> After removing `silent: true`, we're getting duplicate error toasts:
> 
>     1. Global handler shows "Network Error"
> 
>     2. Our `onError` shows "Failed to create appointment"
> 
> 
> i have thought of this approach: Using `toast.dismiss()` in `onError` to dismiss the global toast, then showing our custom message.
> 
> Is this acceptable, or is there a preferred pattern for custom error messages without duplicates?
> 
> onError: (_error) => { toast.dismiss(); // Dismiss global error handler toast toast.error(t("failed_to_create_appointment")); navigate("/patient/home", { replace: true }); },
> 
> And i have addressed all the changes u suggested i will commit them with this one after your suggestion

No don't dismiss the toast, it's fine as is.

harsh04044 committed:
fix: address maintainer feedback for PR #14981

- Remove silent: true from createAppointment mutation
- Remove console.error from onError handler
- Revert unrelated UserForm.tsx changes (buttonLabel refactoring)
- Keep duplicate toasts as is (no toast.dismiss per maintainer feedback)

Addresses all maintainer suggestions on PR #14981

harsh04044 committed:
Merge remote-tracking branch 'upstream/develop' into fix/patient-registration-premature-success-14954

harsh04044 commented:
@Jacobjeevan  i have addressed  your suggestions.

harsh04044 commented:
@Jacobjeevan  review this i have addressed your suggestions , i dont know why changes required tag is not removed.

nihal467 committed:
Merge branch 'develop' into fix/patient-registration-premature-success-14954
